liblumen_eir_interpreter/src/call_result.rs

```diff
@@ -0,0 +1,225 @@
+use core::ptr::NonNull;
+
+use std::convert::TryInto;
+use std::sync::mpsc::{channel, Receiver, Sender};
+use std::sync::Arc;
+
+use liblumen_alloc::borrow::clone_to_process::CloneToProcess;
+use liblumen_alloc::erts::exception::runtime;
+use liblumen_alloc::erts::process::code;
+use liblumen_alloc::erts::process::{Process, Status};
+use liblumen_alloc::erts::term::resource::Reference as ResourceReference;
+use liblumen_alloc::erts::term::{Atom, Boxed, Closure, Term, TypedTerm};
+use liblumen_alloc::erts::{HeapFragment, ModuleFunctionArity};
+
+use lumen_runtime::process::spawn::options::Options;
+use lumen_runtime::scheduler::Scheduler;
+use lumen_runtime::system;
+
+/// A sort of ghetto-future used to get the result from a process
+/// spawn.
+pub struct ProcessResultReceiver {
+    pub process: Arc<Process>,
+    rx: Receiver<ProcessResult>,
+}
+
+impl ProcessResultReceiver {
+    pub fn try_get(&self) -> Option<ProcessResult> {
+        self.rx.try_recv().ok()
+    }
+}
+
+pub struct ProcessResult {
+    pub heap: NonNull<HeapFragment>,
+    pub result: Result<Term, (Term, Term, Term)>,
+}
+
+struct ProcessResultSender {
+    tx: Sender<ProcessResult>,
+}
+
+pub fn call_run_erlang(
+    proc: Arc<Process>,
+    module: Atom,
+    function: Atom,
+    args: &[Term],
+) -> ProcessResult {
+    let recv = call_erlang(proc, module, function, args);
+    let run_arc_process = recv.process.clone();
+
+    loop {
+        let ran = Scheduler::current().run_through(&run_arc_process);
+
+        match *run_arc_process.status.read() {
+            Status::Exiting(_) => {
+                return recv.try_get().unwrap();
+            }
+            Status::Waiting => {
+                if ran {
+                    system::io::puts(&format!(
+                        "WAITING Run queues len = {:?}",
+                        Scheduler::current().run_queues_len()
+                    ));
+                } else {
+                    panic!(
+                        "{:?} did not run.  Deadlock likely in {:#?}",
+                        run_arc_process,
+                        Scheduler::current()
+                    );
+                }
+            }
+            Status::Runnable => {
+                system::io::puts(&format!(
+                    "RUNNABLE Run queues len = {:?}",
+                    Scheduler::current().run_queues_len()
+                ));
+            }
+            Status::Running => {
+                system::io::puts(&format!(
+                    "RUNNING Run queues len = {:?}",
+                    Scheduler::current().run_queues_len()
+                ));
+            }
+        }
+    }
+}
+
+pub fn call_erlang(
+    proc: Arc<Process>,
+    module: Atom,
+    function: Atom,
+    args: &[Term],
+) -> ProcessResultReceiver {
+    let (tx, rx) = channel();
```

Comment:
This is the future-like system you were talking about?

Reply:
That's right. The caveat is that for it to be proper promise, the scheduler loop would have to behave a bit like a tokio reactor. As is, the `call_run_erlang` function alternates between calling the scheduler and polling the future, and then just returns once the result is available.